core/scripts/chaincli/handler/keeper_upkeep_history.go

```diff
@@ -0,0 +1,173 @@
+	var reqs []rpc.BatchElem
+	var results []*string
+	var keeperPerBlock []uint64
+	for block := from; block <= to; block++ {
+		keeperIndex := (uint64(positioningConstant) + ((block - (block % blockCountPerTurn)) / blockCountPerTurn)) % uint64(len(keepersList))
+		payload, err := keeper.RegistryABI.Pack("checkUpkeep", big.NewInt(upkeepId), keepersList[keeperIndex])
+		if err != nil {
+			log.Fatal("failed to pack checkUpkeep: ", err)
+		}
+
+		var result string
+		reqs = append(reqs, rpc.BatchElem{
+			Method: "eth_call",
+			Args: []interface{}{
+				map[string]interface{}{
+					"to":   k.cfg.RegistryAddress,
+					"data": hexutil.Bytes(payload),
+				},
+				// The block at which we want to inspect the upkeep state
+				hexutil.EncodeUint64(block),
+			},
+			Result: &result,
+		})
+
+		results = append(results, &result)
+		keeperPerBlock = append(keeperPerBlock, keeperIndex)
+	}
+
+	log.Println("Doing batch call to check upkeeps")
+	if err := k.rpcClient.BatchCallContext(ctx, reqs); err != nil {
+		log.Fatal("failed to batch call checkUpkeep: ", err)
+	}
+
+	log.Println("Parsing batch call response")
+	type result struct {
+		block          uint64
+		checkUpkeep    bool
+		keeperIndex    uint64
+		keeperAddress  common.Address
+		reason         string
+		performData    string
+		maxLinkPayment *big.Int
+		gasLimit       *big.Int
+		adjustedGasWei *big.Int
+		linkEth        *big.Int
+	}
+	var parsedResults []result
+	for i, req := range reqs {
+		if req.Error != nil {
+			parsedResults = append(parsedResults, result{
+				block:         uint64(i) + from,
+				checkUpkeep:   false,
+				keeperIndex:   keeperPerBlock[i],
+				keeperAddress: keepersList[keeperPerBlock[i]],
+				reason:        strings.TrimPrefix(req.Error.Error(), "execution reverted: "),
+			})
+			continue
+		}
+
+		returnValues, err := keeper.RegistryABI.Methods["checkUpkeep"].
+			Outputs.UnpackValues(hexutil.MustDecode(*results[i]))
+		if err != nil {
+			log.Fatal("unpack checkUpkeep return: ", err, *results[i])
+		}
+
+		parsedResults = append(parsedResults, result{
+			block:          uint64(i) + from,
+			checkUpkeep:    true,
+			keeperIndex:    keeperPerBlock[i],
+			keeperAddress:  keepersList[keeperPerBlock[i]],
+			performData:    "0x" + hex.EncodeToString(*abi.ConvertType(returnValues[0], new([]byte)).(*[]byte)),
+			maxLinkPayment: *abi.ConvertType(returnValues[1], new(*big.Int)).(**big.Int),
+			gasLimit:       *abi.ConvertType(returnValues[2], new(*big.Int)).(**big.Int),
+			adjustedGasWei: *abi.ConvertType(returnValues[3], new(*big.Int)).(**big.Int),
+			linkEth:        *abi.ConvertType(returnValues[4], new(*big.Int)).(**big.Int),
+		})
+	}
+
+	writer := tabwriter.NewWriter(os.Stdout, 8, 8, 0, '\t', 0)
+	defer writer.Flush()
+
+	fmt.Fprintf(writer, "\n %s\t\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t", "Block", "checkUpkeep", "Keeper Index", "Keeper Address", "Max LINK Payment", "Gas Limit", "Adjusted Gas", "LINK ETH", "Perform Data", "Reason")
```

Comment:
I didn't realize you did all this by hand! Is there not like a "table printer" package we could use? Just thinking about the complexity of adding new rows, etc...

Reply:
Good point. My idea is to avoid unnecessary dependencies for simple logic. Let's keep it without an extra package until this logic is became more complex.